### PR TITLE
Added documentation inline

### PR DESCRIPTION
**Pull request by stevekuznetsov**
Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

@danmcp I'll add a comment tagging you when this is ready to look at

### CODE REVIEW

File: test_pull_requests [hunk 1]
```diff
@@ -1106,6 +1139,18 @@ popd
 
     end
 
+    # The GitHub mergeable API is flaky, so we use an external file ``database'' to record
+    # pull request mergeability responses with a thresholded saturating counter so that pull
+    # request comment/label actions are only taken after a string of consequtive API responses.
```
Comment (danmcp) [inline test_pull_requests]: Spelling: consecutive

But it's not really consecutive.  It's totaling.

File: test_pull_requests [hunk 2]
```diff
@@ -1106,6 +1139,18 @@ popd
...
+    # The current thresholds are at 0 and 10, with an initial state of 0, so i.e. the default
+    # state is to consider the pull request mergeable, and this state will not change until
+    # a net of ten unmergeable responses come in from the GitHub API. When set_mergeable is
```
Comment (danmcp) [inline test_pull_requests]: It's not that the state won't change.  One not mergeable and we immediately won't try to test it.  We just wait to add the comment/label until we are sure.

File: test_pull_requests [hunk 3]
```diff
@@ -1132,12 +1193,16 @@ popd
...
+        # if our counter is saturated at 10, we want to add the comment and label
         create_or_update_comment(id, repo, ACTION_PREFIX, ACTION_NOT_MERGE)
         ensure_labels(id, repo, [NEEDS_REBASE_LABEL])
       else
+        # if our counter is not saturated, we pull it towards 10
         `sed -i "/#{merge_id}=/d" ~/test_pull_request_not_mergeable && echo "#{merge_id}=#{(count+1).to_s}" >> ~/test_pull_request_not_mergeable`
       end
-      count #return count of how many times this has happened (used as stability measure, Github becomes unstable for mergeability right after a merge happens)
+      # return count of how many times this has happened, for use as a stability measure,
+      # Github becomes unstable for mergeability right after a merge happens)
```
Comment (danmcp) [inline test_pull_requests]: Github should always be GitHub

Comment (danmcp) [inline test_pull_requests]: too many commas /  not a sentence

File: test_pull_requests (hunk 1 above)
Comment (stevekuznetsov) [inline test_pull_requests]: I swear I can English well...

**Comment by stevekuznetsov** [comment]
@danmcp comment away

**Comment by stevekuznetsov** [comment]
`TODO` is contentious... I expect you to comment on or reject a lot of them

File: test_pull_requests [hunk 4]
```diff
@@ -649,15 +668,12 @@ popd
       return comment
     end
 
-    #
-    # Gets a comment with given prefix
-    #
+    # get_comment_with_prefix returns a bot comment with the given prefix, if one exists
```
Comment (danmcp) [inline test_pull_requests]: returns a comment by the bot with ...

Bot comment means a very specific comment.

File: test_pull_requests [hunk 5]
```diff
@@ -803,6 +843,7 @@ popd
     end
 
     # Updates all the bot markers for the involved pull requests.  To be called after you have validated the tests should be kicked off again
+    # update_evaluated_markers
```
Comment (danmcp) [inline test_pull_requests]: this is consistent with the rest of the methods

File: test_pull_requests [hunk 6]
```diff
@@ -827,6 +868,9 @@ popd
       end
     end
 
+    # get_updated_at returns the time at which the last commit on the pull request was created, and
+    # whether or not the pull request has been updated since the last time the bot saw it, determined by
+    # looking at the commit referenced in the last bot comment and comparing the latest commit SHA
```
Comment (danmcp) [inline test_pull_requests]: in the last evaluated comment

File: test_pull_requests (hunk 5 above)
Comment (stevekuznetsov) [inline test_pull_requests]: can remove, or can flesh out comment... preference?

File: test_pull_requests [hunk 7]
```diff
@@ -854,14 +898,22 @@ popd
...
+      trigger_author_trusted = user_trusted?(trigger_login, addtl_pull_repo, settings)
+      if author_trusted || (trigger_author_trusted && (trigger_updated_at && trigger_updated_at >= updated_at))
         pull_request = get_pull_request(addtl_pull_id, addtl_pull_repo, 2)
         if pull_request
-          if pull_request['mergeable']
+          # on co-requisite pull requests, we do not expect there to be trigger statements, as the 
+          # triggers on the parent pull request will trigger builds and tests with the co-requisites
+          # therefore, we need to manage the mergeable state of the co-requisite pull requests here
+          # while we're considering them for their parent
+          if is_mergeable?(base_pull_id, base_repo)
```
Comment (stevekuznetsov) [inline test_pull_requests]: non-comment change

File: test_pull_requests [hunk 8]
```diff
@@ -854,14 +898,22 @@ popd
...
+    # is trusted and the co-requisite pull request is merge-able.
     def add_coreq(addtl_pull_id, addtl_pull_repo, login, trigger_login, repo_to_pull_request, settings, trigger_updated_at, updated_at, base_pull_id, base_repo, comments)
       $stderr.puts "  Processing dependency on pull request #{addtl_pull_id} in repo '#{addtl_pull_repo}'"
-      trusted = user_trusted?(login, addtl_pull_repo, settings)
-      # It's ok if the user is trusted or the trigger was added after the comment
-      if trusted || (user_trusted?(trigger_login, addtl_pull_repo, settings) && (trigger_updated_at && trigger_updated_at >= updated_at))
+      # we can add the co-requisite if the author of the original pull request is trusted, or if the author of the trigger
+      # statement is trusted and they authored the trigger statement after the co-requisite statement was made
+      author_trusted = user_trusted?(login, addtl_pull_repo, settings)
+      trigger_author_trusted = user_trusted?(trigger_login, addtl_pull_repo, settings)
```
Comment (stevekuznetsov) [inline test_pull_requests]: non-comment change

File: test_pull_requests [hunk 9]
```diff
@@ -948,15 +1013,39 @@ popd
                   break
                 end
               end
+              # the main pull request and all of the co-requisite pull requests haven't been
+              # updated since the last trusted trigger, so we can begin testing
               update_evaluated_markers(repo_to_pull_request, trigger_updated_at, settings)
               submit_test_job = true
             else
+              # no trusted trigger statement has been made, so we cannot build or test this pull
+              # TODO: only pull requests with trusted triggers get process_pull_request called
```
Comment (danmcp) [inline test_pull_requests]: I am not sure this is true.  ppr is called when you know you have a comment from a trusted user.  But I don't think we know if that comment is at a place we can consider it a trusted trigger yet.

File: test_pull_requests [hunk 10]
```diff
@@ -1006,11 +1103,37 @@ popd
             status = result == 'SUCCESS' ? 'success' : 'failure'
           end
         else
+          # in this case, we're in one of three states:
```
Comment (danmcp) [inline test_pull_requests]: You are inconsistent with capitalization.  This starts with a lower case,  but the below paragraph starts with upper.

File: test_pull_requests [hunk 11]
```diff
@@ -1006,11 +1103,37 @@ popd
...
+          #  3) the build has finished, with one of the following states:
+          #    - SUCCESS
+          #    - FAILURE
+          #    - ABORTED
+          #    - UNSTABLE
+          #    - NOTFOUND
+          #
+          # Regardless of the current state, since we have a trusted trigger, we want to
+          # re-submit this pull request for testing.
+          # TODO: technically, in states 1) and 2), we don't necessarily know if it's ever
```
Comment (danmcp) [inline test_pull_requests]: What's the harm in resubmit?  Here or the case you listed above.  Not sure it's worth the comments.

File: test_pull_requests [hunk 12]
```diff
@@ -1061,21 +1184,36 @@ popd
...
               update_status(settings['test_prefix'], commit['sha'], repo, status, build_url, (status == 'success') ? 'Passed' : 'Failed')
             end
             next if repo == base_repo
             recreate_comment_with_prefix(pull_request['number'], repo, settings['test_prefix'], updated_comment)
           end
         else
+          # if we are not running tests and have not specified a literal state to transition
+          # into, we have one more transition to check: from post-build sucess to merge
+          #
+          # TODO: this seems out of place
```
Comment (danmcp) [inline test_pull_requests]: Remove this one.  Too generic.

File: test_pull_requests [hunk 13]
```diff
@@ -1106,6 +1244,22 @@ popd
...
+    # request comment/label actions are only taken after a critical amount of net positive
+    # or negative API responses. However, a pull request is immediately chosen not to be 
+    # tested after the first unmergeable result, even if it takes longer for the UX comment
+    # and label to be applied.
+    #
+    # The saturation points are at 0 and 10, with an initial state of 0, so i.e. the default
+    # state is to consider the pull request mergeable, and this state will not change until
+    # a net of ten unmergeable responses come in from the GitHub API. When set_mergeable is
+    # called, the counter for the specific pull ID is decremented if it hasn't saturated; 
+    # when set_not_mergeable is called, the counter is incremented unless its saturated.
```
Comment (danmcp) [inline test_pull_requests]: it's

File: test_pull_requests [hunk 14]
```diff
@@ -1155,9 +1328,20 @@ popd
...
           if $branches.include?(branch) || $branches.include?('*')
 
             $stderr.puts "  Updated at: #{req['updated_at']}"
+            # we only want to consider pull requests that have been modified in
+            # the last twelve hours, to stop us from doing extra work when we
+            # don't need to. Also, just to ensure that we don't forget a pull
+            # request forever on accident, there is a 10% chance we'll consider
+            # a pull request even if it is inactive
+            #
+            # TODO: wouldn't this be better if we brought this time down to
```
Comment (danmcp) [inline test_pull_requests]: Remove.  Way too specific for this repo which is more generic than the case you are talking about.

File: test_pull_requests [hunk 15]
```diff
@@ -1169,13 +1353,19 @@ popd
                 set_mergeable(id, repo, login, comments)
               else
                 if set_not_mergeable(id, repo, login) == 0
+                  # TODO: since our counter saturates at 0, why would we ever
```
Comment (danmcp) [inline test_pull_requests]: This is the previous value.

File: test_pull_requests [hunk 16]
```diff
@@ -1169,13 +1353,19 @@ popd
...
                 end
                 next
               end
 
               comments = get_comments(id, repo) if comments.nil?
 
+              # we only want to consider pull requests where the last trigger we found
+              # is from a trusted user
+              #
+              # TODO: use trusted_trigger_time?
```
Comment (danmcp) [inline test_pull_requests]: We have to get further in the logic to update the PR with the appropriate messages if the pull request isn't appropriate to test.

File: test_pull_requests [hunk 17]
```diff
@@ -1241,6 +1433,11 @@ popd
         settings = req_info[4]
         branch = req['base']['ref']
 
+        # if we're only allowing sequential tests in this tag, we want to find
+        # any pull request in the 'running tests' state and signal that there
+        # is a test running
+        #
+        # TODO: seems out of place
```
Comment (danmcp) [inline test_pull_requests]: Too generic.

File: test_pull_requests (hunk 5 above)
Comment (danmcp) [inline test_pull_requests]: To be clear what I meant.  The rest of the method comments start with the method name and then describe what they do.  This one is the opposite.

File: test_pull_requests (hunk 7 above)
Comment (danmcp) [inline test_pull_requests]: This looks like you are checking against the wrong pull.  I am not sure if this is correct yet.  But can you move it to a separate pull to debate.

File: test_pull_requests (hunk 15 above)
Comment (stevekuznetsov) [inline test_pull_requests]: that's surprising

File: test_pull_requests (hunk 10 above)
Comment (stevekuznetsov) [inline test_pull_requests]: what would you prefer? my insane reasoning is comments don't start with capitals, but after a period a sentence should be capital... should I just capitalize everything?

Comment (danmcp) [inline test_pull_requests]: I prefer start with capital.  That seems the most consistent overall.

File: test_pull_requests (hunk 4 above)
Comment (stevekuznetsov) [inline test_pull_requests]: > Bot comment means a very specific comment.

unclear what you mean -- in this context this is not a special comment, just one with the prefix

Comment (stevekuznetsov) [inline test_pull_requests]: oh, you mean `'bot comment'` === `'bot comment about evaluating things'`

File: test_pull_requests (hunk 9 above)
Comment (stevekuznetsov) [inline test_pull_requests]: there's probably some nuance I'm missing, but why, exactly, can't we replace [THIS](https://github.com/stevekuznetsov/test-pull-requests/blob/master/test_pull_requests#L1174-L1203) block with `trusted_trigger_time`? We seem to update with the `ACTION_NOT_TEAM` comment right then and there if we don't find a trusted trigger

File: test_pull_requests (hunk 11 above)
Comment (stevekuznetsov) [inline test_pull_requests]: makes it harder to think about what state the machine is in when you see `resubmit` when in reality it's a first submission

Comment (stevekuznetsov) [inline test_pull_requests]: but no harm

File: test_pull_requests [hunk 18]
```diff
@@ -854,13 +901,21 @@ popd
...
+    # is trusted and the co-requisite pull request is merge-able.
     def add_coreq(addtl_pull_id, addtl_pull_repo, login, trigger_login, repo_to_pull_request, settings, trigger_updated_at, updated_at, base_pull_id, base_repo, comments)
       $stderr.puts "  Processing dependency on pull request #{addtl_pull_id} in repo '#{addtl_pull_repo}'"
-      trusted = user_trusted?(login, addtl_pull_repo, settings)
-      # It's ok if the user is trusted or the trigger was added after the comment
-      if trusted || (user_trusted?(trigger_login, addtl_pull_repo, settings) && (trigger_updated_at && trigger_updated_at >= updated_at))
+      # We can add the co-requisite if the author of the original pull request is trusted, or if the author of the trigger
+      # statement is trusted and they authored the trigger statement after the co-requisite statement was made
+      author_trusted = user_trusted?(login, addtl_pull_repo, settings)
+      trigger_author_trusted = user_trusted?(trigger_login, addtl_pull_repo, settings)
```
Comment (stevekuznetsov) [inline test_pull_requests]: non-comment diff

**Comment by stevekuznetsov** [comment]
@danmcp responded to comments

File: test_pull_requests (hunk 18 above)
Comment (danmcp) [inline test_pull_requests]: I got it.  It's good.

File: test_pull_requests (hunk 9 above)
Comment (danmcp) [inline test_pull_requests]: This logic is not getting the list of commits I think is the main difference so as to save the api calls.  There is certainly some logic that could be shared though.

**Comment by stevekuznetsov** [comment]
@danmcp any final comments?

File: test_pull_requests [hunk 19]
```diff
@@ -1106,6 +1239,24 @@ popd
...
+    # tested after the first unmergeable result, even if it takes longer for the UX comment
+    # and label to be applied.
+    #
+    # The saturation points are at 0 and 10, with an initial state of 0, so i.e. the default
+    # state is to consider the pull request mergeable, and this state will not change until
+    # a net of four unmergeable responses come in from the GitHub API. When set_mergeable is
+    # called, the counter for the specific pull ID is decremented if it hasn't saturated; 
+    # when set_not_mergeable is called, the counter is incremented unless it is saturated.
+    #
+    # TODO: if we saturate at 10, why do we threshold at 4?
```
Comment (danmcp) [inline test_pull_requests]: Are you talking about for the coreq case?  I think that's a bug.  Should be sharing a constant.

Comment (stevekuznetsov) [inline test_pull_requests]: yes, I'll remove this TODO and make an issue

**Comment by stevekuznetsov** [comment]
@danmcp squashed, added constants

File: test_pull_requests [hunk 20]
```diff
@@ -1106,38 +1245,77 @@ popd
...
         ensure_labels(id, repo, [NEEDS_REBASE_LABEL])
       else
+        # If our counter is not saturated, we pull it towards 10
         `sed -i "/#{merge_id}=/d" ~/test_pull_request_not_mergeable && echo "#{merge_id}=#{(count+1).to_s}" >> ~/test_pull_request_not_mergeable`
       end
-      count #return count of how many times this has happened (used as stability measure, Github becomes unstable for mergeability right after a merge happens)
+      # return the previous value of the saturating counter so callers can consume it to
+      # determine stability level
+      #
+      # TODO: return the current value of the counter, since it gives a better glimpse
```
Comment (danmcp) [inline test_pull_requests]: Remove please.  It's missing the point of what this is for.

**Comment by stevekuznetsov** [comment]
@danmcp addressed comment

**Comment by danmcp** [comment]
@stevekuznetsov Needs rebase

**Comment by stevekuznetsov** [comment]
@danmcp rebased

**Pull request merged**